### PR TITLE
fix(gpws): fixed mode 4 logic

### LINKED ISSUE #8134: GPWS Mode 4 behaviour
### Aircraft Version

Development

### Build info

```json
{
    "built": "2023-07-23T18:08:26+00:00",
    "ref": "refs/heads/master",
    "sha": "5a800afc6f9127acf5ff0971f58d69fd54732cfc",
    "actor": "BlueberryKing",
    "event_name": "manual",
    "pretty_release_name": "master:5a800afc",
    "version": "v0.11.0-dev.5a800af"
}
```

### Describe the bug

It is virtually impossible to activate Mode 4A or Mode 4B warnings specifically the "Too low flaps" and "Too low gear" aural warnings. I belive there is an oversight in the code which is not allowing the Mode 4A/B aural warnings to be executed since it is always overridden by the Mode 4C therefore even though the aircraft is in the envelope for 4A or 4B activation, only the 4C "Too low terrain" aural warning is heard. 

### Expected behavior

I belive that the Mode 4C should only work during TO and initial climb and should not trigger a warning after passing 2400ft RA. After this point, the functionality of the Mode 4C should be handled by the high speed envelope of the Mode 4A and Mode 4B. During standard flight profile in the current implementation you will only recive the Mode 4C warning.

### Steps to reproduce

A) Perform standard approach with LG UP.
- Observe as Mode 4C triggers a "Too low terrain" aural warning after passing 500ft RA. "Too low gear" should be heard instead.

B) Perform standard approach with Flaps not in landing config.
- Observe as Mode 4C triggers a "Too low terrain" aural warning after passing 500ft RA and continues to repeat this warning even after passing 245ft (The upper limit of Mode 4B envelope) where the "Too low flaps" should be heard instead.

### References (optional)

98028460-6d79b400-1e0e-11eb-975f-934d29067444
98028535-88e4bf00-1e0e-11eb-90d3-9f5ec0435f59

These pictures are from the A320 FCOM.

### Additional info (optional)

_No response_

### Discord Username (optional)

leech2188

### PR DESCRIPTION
**Pull request by LeechCZ**
Fixes #8134

## Summary of Changes

Mode 4C logic has been changed so it can only be triggered during take-off and go-around and does not trigger warnings during approaches.
Mode 4A and Mode 4B logic has been changed so it can be triggered even when the FMGC is not in the Approach phase.
The general Mode 4 logic has been changed so that the Mode 4C can only be triggered before passing 2400 ft RA during a take-off or go-around. After that point, only Mode 4A and 4B warnings can be triggered to be available during an approach.
## References

98028460-6d79b400-1e0e-11eb-975f-934d29067444
98028535-88e4bf00-1e0e-11eb-90d3-9f5ec0435f59
These pictures are from the A320 FCOM.
## Additional context

Discord username (if different from GitHub): leech2188

## Testing instructions

1. Check for normal GPWS operation during a flight (mainly no erroneous Mode 4 warnings, other modes should not be affected)
2. Test for normal Mode 4 warnings as shown in References:
    - Mode 4C warnings during a take-off or go-around
    - Mode 4A / 4B warnings during an approach

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Review by 2hwk — APPROVED** [review]

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_GPWS.js [hunk 1]
```diff
@@ -179,11 +186,34 @@ class A32NX_GPWS {
     }
 
     update_maxRA(radioAlt, onGround, phase) {
+
+
```
Comment (tracernz) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_GPWS.js:190]:
```suggestion
```

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_GPWS.js [hunk 2]
```diff
@@ -364,36 +394,38 @@ class A32NX_GPWS {
      * @param phase - Flight phase index
      * @constructor
      */
-    GPWSMode4(mode, radioAlt, speed, FlapsInLandingConfig, gearExtended, phase) {
+    GPWSMode4(mode, radioAlt, speed, FlapsInLandingConfig, gearExtended) {
+
```
Comment (tracernz) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_GPWS.js:398]:
```suggestion
```

**Review by tracernz — CHANGES_REQUESTED** [review]
Code style fixes.

**Review by tracernz — APPROVED** [review]
Formatting is good now. I haven't reviewed the logic against my refs at all.

**Review by BlueberryKing — DISMISSED** [review]
Thanks for the PR! Apart from my suggestions, there is a small issue with the logic. Mode 4A is only meant to trigger if the gear is up and the flaps are not in landing configuration. Then, Mode 4B can also issue "TOO LOW GEAR" in the envelope where it would issue "TOO LOW FLAPS" if the gear is not down.

**Comment by LeechCZ** [comment]
> Thanks for the PR! Apart from my suggestions, there is a small issue with the logic. Mode 4A is only meant to trigger if the gear is up and the flaps are not in landing configuration. Then, Mode 4B can also issue "TOO LOW GEAR" in the envelope where it would issue "TOO LOW FLAPS" if the gear is not down.

If you have any other refference, could you please provide it to me? As I understand it, from the FCOM provided in the PR, Mode 4A can only be triggered if the LG is not down. And the Mode 4B can only be triggered if the LG is down and the flaps are not in landing configuration. Althoug it also says in the FCOM that Mode 4B can trigger the "TOO LOW GEAR" warning, but that is at least to my understanding contradictory to the trigger conditions mentioned above in the document.

**Comment by beheh** [comment]
@2hwk are you sure this RTT? Looking at the comments this PR seems to have some deficiencies that still need to be addressed.

**Review by beheh — CHANGES_REQUESTED** [review]
Flagging this PR as "changes requested" due to the mentioned logic from the most recent comments.

**Comment by BlueberryKing** [comment]
After discussing with the author, I've addressed the suggestions, so the PR can be re-reviewed now.

**Review by 2hwk — DISMISSED** [review]

**Comment by noonebling** [comment]
Quality Assurance Trainee Report

Discord          : noonebling
Object of testing: 8136
Tier of Testing  : 1
Date             : 31/05/2024

Testing Process: 
Checked for normal GPWS operation during a flight
Test for normal Mode 4 warning
1. Mode 4C during take off
2. Mode 4A/4B warnings during an approach

Negatives:
N/A

Testing Results: 
Passed

Conclusions: 
Works as expected

**Comment by beheh** [comment]
I find it fairly suspicious that is using FMGC phases now, as generally I believe the GPWS is standalone and should not really depend on the aircraft inputting certain things (imagine an FMGC failure, you'd still need the GPWS to work properly!). How sure are we that the takeoff/landing detection is not done via RA thresholds and/or timings?

**Comment by beheh** [comment]
We've FWC flight phases for a long, long time :-) They're in an LVar, shouldn't be that hard.

**Comment by lukecologne** [comment]
if it's using any flight phase, it's the FMGC one. The GPWC doesn't even have an input bus from the FWC

**Comment by LeechCZ** [comment]
The way my rework was intended was to not use flight phases at all since to my knowledge that is not how the system operates. Instead my intention was the dertemine the "phase" and envelopes for the warnings using RA. I will try to look it over in the comming week or at least give something to BlueberryKing since I'm really swamped rn.

**Comment by Al-Roker** [comment]
I believe this excerpt provides the method EGPWS uses to select the current phase of flight.

AirbusSwitching

**Comment by beheh** [comment]
Having the GPWS determine this state makes a lot more sense! That description makes it sound like switching can be done based on the RA an gear/flap status alone, so it shouldn't be too hard.

**Review by beheh — APPROVED** [review]
I'm happy with the code, but have not done an accuracy check.

**Comment by Aleloplt** [comment]
Quality Assurance Tester/Trainee Report

Discord          : Alelo
Object of testing: 8136
Tier of Testing  : (1/2)
Date             : 31/07/2024

Testing Process: 
Initial flight from EGLL - EGFF. 
Full start up to pushback in normal time
Approach into EGFF and multiple go arounds to test different modes
First Test - Mode 4C 
- AP disconnected
- TOGA thrust engaged
- Pitch reduced to minimise height gain - steady 100'RA flight
- "too low terrain audible"
- Pitched up to gain altitude and eventually GPWS call out disappeared
-*IGNORE VFE OVSPD* - i did this myself

Second Test - Mode 4A
- Normal approach w/ AP on right down to MNM.
- Flaps full, LDG gear up, VAPP 140kn
- "too low gear" audible upon passing 490'RA
- Go around and tested again - including mode 4C test
- Descent towards terrain using FPA -3.0 at 195knots
- "Too low Terrain" only audible roughly at 600'RA

Third Test - Mode 4B
- Normal approach using FPA -3.0
- LDG down, Flap 3 160 knots
- "too low terrain" audible 250'RA
- "too low flaps" audible instantly at 240'RA
- 
Negatives:       
The exact call outs dont seem to match the FCOM document provided. I.e should be hearing too low terrain less than 1000'RA without the gear and at 196knots but that was coming in far later and im not sure if thats meant to be correct

Testing Results: 
Mode 4C - **Passed**
Mode 4A - Gear call out - **Passed**
Mode 4B - Flaps call - **Passed**

Conclusions: 
need more clarity on the 'too low terrain calls and whether not hearing them as per the FCOM provided is allowable

Media:       
Media files are too big to be sent here. If you PM me on discord ill send them across

**Comment by beheh** [comment]
Thanks for the QA! Merging this is then just pending an accuracy check.

**Pull request merged**